pkg/adapter/awscognitouserpoolsource/adapter_test.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright (c) 2019-2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awscognitouserpoolsource
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider/cognitoidentityprovideriface"
+
+	adaptertest "knative.dev/eventing/pkg/adapter/v2/test"
+	loggingtesting "knative.dev/pkg/logging/testing"
+)
+
+type mockedCognitoUserPoolClient struct {
+	cognitoidentityprovideriface.CognitoIdentityProviderAPI
+	listUsersOutput cognitoidentityprovider.ListUsersOutput
+}
+
+func (m mockedCognitoUserPoolClient) ListUsers(in *cognitoidentityprovider.ListUsersInput) (*cognitoidentityprovider.ListUsersOutput, error) {
+	return &m.listUsersOutput, nil
+}
+
+func TestListUsers(t *testing.T) {
+	user := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now()
+			return &now
+		}(),
+	}
+
+	a := &adapter{
+		userPoolID: "userpool/fooPool",
+		logger:     loggingtesting.TestLogger(t),
+		ceClient:   adaptertest.NewTestClient(),
+		cgnIdentityClient: mockedCognitoUserPoolClient{
+			listUsersOutput: cognitoidentityprovider.ListUsersOutput{
+				Users: []*cognitoidentityprovider.UserType{
+					&user,
+				},
+			},
+		},
+	}
+
+	users, err := a.listUsers()
+	assert.NoError(t, err)
+	assert.Equal(t, []*cognitoidentityprovider.UserType{&user}, users)
+}
+
+func TestFilterByTimestamp(t *testing.T) {
+	now := time.Now()
+
+	// existing user
+	user1 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: &now,
+	}
+
+	// we must get 0 new users
+	users, timestamp := filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+	}, now)
+	assert.Len(t, users, 0)
+
+	// new user
+	user2 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: func() *time.Time {
+			now := time.Now()
```

Comment:
nit: This works because the code is sequential and the time is down to the nanosecond, but as a general practice in tests you could also fake timestamps with [`time.Unix(0, 0)`](https://golang.org/pkg/time/#Unix) (and increment manually) to ensure they work in any situation.